### ISSUE TITLE
Dismissable "Images" status warning

### ISSUE THREAD

**Issue by bugfolder** [backdrop-contrib/ubercart#436]
On the Status report at `admin/reports/status`, if one has not let Ubercart automatically configure the image file field, one sees this warning:

>[Automatically configure the following items for core image support] (links to: http://mysite/admin/store/settings/products/defaults):
>The Image file field has not been created for products.
>(This action is not required and should not be taken if you do not need images or have implemented your own image support.)

If one has implemented their own image support, there is no need to display a constant warning on the status page, particular since users typically want to see "all green" on the status page. So this warning should be dismissable. For example, by add a checkbox to the product config settings, e.g., "Don't warn about automatic image configuration on status page."

Or perhaps provide a way to set `'uc_image_product'` in `uc_product.settings` to the appropriate field, or `--None--` if the user does not need images. This config value was automatically set in `uc_product_update_1003()`, but there's no ability to edit it to something else from within the UI.

**Comment by bugfolder** [comment]
Related to this: on the Product settings page, `admin/store/settings/products`, there is a setting for the Product image widget:

![product image](https://user-images.githubusercontent.com/29214084/224088123-45b5c168-6092-4cea-8d1d-f65b55bc8aaa.jpg)

This widget is only used if the user selects the display format "Ubercart product", which could be used with the default Ubercart image field `'uc_product_image'`or a custom image field (possibly inherited from a D7 upgrade). So that should be made clearer in the help text for that setting.

**Comment by bugfolder** [comment]
There's a lot to complain about with the status report warning that is shown if you've upgraded from D7 but don't use the built-in Ubercart image field. It looks like this:

![status](https://user-images.githubusercontent.com/29214084/224192295-f8cb962d-f9d4-4f3d-96f5-8008defff915.jpg)

* The generic "Images" title looks like something that refers to site-wide images, when this is really just about Ubercart.
* The "Automatically configure the following items..." link does not make clear that this is actually a dangerous link; one would think that clicking it would take you to the page where configuration is, um, configured, but in fact, it goes ahead and creates new fields on all product types, somethng that is not easily undoable (especially if you have multiple product types).
* and then it dumps you at `admin/store`, which is entirely different from where you were, so you don't see the result of what you just did.

And in general, one should not be making configuration settings from the Status report; one should do that sort of thing on a Settings page. The status report should just tell you what the status of things is and then send you to the place where you would make associated settings changes.

So I've moved the selection of functionality to the Product settings page, made the Status page just report status, but also provide links to the Product Settings page, if appropriate.

The Product Settings page now looks like this. 

![product settings](https://user-images.githubusercontent.com/29214084/224193081-0508515f-6564-4dd1-a4fb-986197ab5b9e.jpg)

I've added two radio buttons: the upper one chooses to use the Ubercart image field and provides that link that configures the image fields (and after configuring, it brings you back to this page). The lower one is the choice to not use the default image field; if this is chosen, you won't get that warning any more on the status page.

And if you've chosen the lower button or you've already configured the core image fields, that "Automatically configure..." link goes away:

![no link](https://user-images.githubusercontent.com/29214084/224193437-38a1610e-3454-4fe2-88f0-dc9073c95932.jpg)

PR to follow.

**Comment by argiepiano** [comment]
@bugfolder, can you post steps to get that warning you are describing? I can't see this, and therefore I can't test the PR.

**Comment by argiepiano** [comment]
Never mind. I figured it out. I'll test this soon.

**Comment by argiepiano** [comment]
Me again. I'm doing an installation of Ubercart from scratch (with this patch) on a clean installation of Backdrop.

So, upon installing Ubercart, without having done anything, I see the following:
- Status report says `Ubercart products are configured to use the default product image field.` (and, after checking, I can see that the Product content type has `uc_product_image` attached). I was expecting that, upon a clean installation, Ubercart would not create the field and attach it to the default product content type (?)
- When I visit `admin/store/settings/products` I see that the option `No images needed or other image support` is checked (?) despite the fact that the status report said otherwise
- When I try to click `Default Ubercart image field` in that page, there is an AJAX call. The radio button is checked for 1 second, and then goes back to the other option, `No images needed or other image support`

I haven't looked into what's causing all these inconsistencies... but I thought I'd report them.

![screencapture](https://user-images.githubusercontent.com/9938978/224223493-6903a17c-fe0e-4aef-8af4-87cfe9892a77.gif)

**Comment by bugfolder** [comment]
>  I was expecting that, upon a clean installation, Ubercart would not create the field and attach it to the default product content type (?)

Auto-creating the field is the default behavior (with or without this patch).

To get to the warning state, install UC, then delete the Ubercart-created **Image** fields from the Product content type.

I got to the warning state via upgrading a D7 site that had its own image fields on products. After cleaning up the content types by removing the superfluous **Image** fields, the warning started appearing.

> When I try to click Default Ubercart image field in that page, there is an AJAX call. The radio button is checked for 1 second, and then goes back to the other option...

That's not supposed to happen! So I'll look into that further.

**Comment by bugfolder** [comment]
Revised the PR. The AJAX issue is (I think) fixed.

The question of whether UC should auto-configure the product image field is mixed into this. I think it makes sense to do so for a fresh install, but not to automatically add the field for upgrades from D7 if it wasn't there in the D7 site.

**Comment by bugfolder** [comment]
Incidentally, in looking into https://github.com/backdrop-contrib/ubercart/issues/440, I see that (aside from that issue's problem), UC File Downloads handles the status report properly (IMHO):

![downloads error](https://user-images.githubusercontent.com/29214084/224377726-abc30010-ffc4-4d71-87a3-99f396ebc355.jpg)

It flags the problem, and then links to the page where you would address it.

**Comment by argiepiano** [comment]
@bugfolder, I'm still seeing the same inconsistency upon a clean install:

- Status report says "Ubercart products are configured to use the default product image field." for Product Images
- `admin/store/settings/products` is showing "No images needed or other image support" as the selected option

These contradict one another, unless I'm not understanding the meaning of this.

**Comment by argiepiano** [comment]
And the AJAX issue is still a problem

**Comment by bugfolder** [comment]
That's a little puzzling. I do a clean minimal install of Ubercart on a bare Backdrop installation, and I see this in the status report:

![status](https://user-images.githubusercontent.com/29214084/224490513-8ace4bea-369e-42cc-a88c-63aa772f7131.jpg)
 and this on the settings:
![settings](https://user-images.githubusercontent.com/29214084/224490520-529e3662-dd7e-4a6f-9e5f-d5177a4fec95.jpg)

And that's correct.

And then I go delete the "Image" field from the Product content type, and I see this for status:

![status](https://user-images.githubusercontent.com/29214084/224490732-b3918995-dd0d-4332-9bce-29528c0c8702.jpg)
 and this for settings:

![settings](https://user-images.githubusercontent.com/29214084/224490738-4008442e-77f7-4cef-b517-fe8a798b7f84.jpg)

and that's correct.

And then, on the settings page, if I switch to the upper radio button, the description changes (via AJAX) to this:

![settings 2](https://user-images.githubusercontent.com/29214084/224490803-b5918309-e741-46c9-b2f0-2b1426eb99d9.jpg)

And if I save that setting, the Status report switches to the warning:
![status warning](https://user-images.githubusercontent.com/29214084/224490865-f76e98c4-2b3f-4283-b9ab-849336bf33c4.jpg)

I'm testing with PHP 7.4.

Can you describe the exact sequence of actions you're doing to get to what you're seeing? (And confirm you're using the latest diff, https://patch-diff.githubusercontent.com/raw/backdrop-contrib/ubercart/pull/439.diff)?

**Comment by argiepiano** [comment]
I'm thinking my installation must have something wrong... I was messing with the codebase for other stuff. Let me try again and I'll report.

**Comment by argiepiano** [comment]
Problem still exists as I described when I use PHP 7.4..33. PR works fine in PHP 8.1 (!).

Can you test in 7.4?

**Comment by bugfolder** [comment]
Ah, my test setup for modules was left in PHP 8.1 mode. Heading back to 7.4...

**Comment by bugfolder** [comment]
OK, this is crazy! In PHP 7.4, the order of the options matters if one of the keys is 0!

Put this into a form and watch it render:

```
  $form['radios_1'] = array(
    '#type' => 'radios',
    '#title' => t('Radios 1'),
    '#options' => array(
      'foo' => t('Foo option'),
      0 => t('Zero option'),
    ),
    '#default_value' => 'foo',
  );
  $form['radios_2'] = array(
    '#type' => 'radios',
    '#title' => t('Radios 2'),
    '#options' => array(
      0 => t('Zero option'),
      'foo' => t('Foo option'),
    ),
    '#default_value' => 'foo',
  );
```
And here's how it renders in PHP 7.4. Note that the wrong default value is selected when the 0 option comes last. It renders correctly in PHP 8.1. And this is why the AJAX is misbehaving.

![radios](https://user-images.githubusercontent.com/29214084/224494491-9981886e-ea58-4177-a411-0f1aa98b5b4e.jpg)

**Comment by bugfolder** [comment]
I think the solution here will be to use an empty string, rather than 0, as the key for "None or other". I'll make that change shortly.

**Comment by bugfolder** [comment]
PR updated to use empty string as the key. Seems to work in both PHP 7.4 and 8.1. (Also gonna file a core issue on the discrepancy.)

**Comment by argiepiano** [comment]
Thanks for all this work, @bugfolder!

In testing, this is working great now, even in 7.4. One suggestion about this text:

`Not configured yet. Automatically configure Ubercart products for core image support.`

I think this vague and slightly confusing and/or prone to user error. As a user, I just clicked the choice "`Default Ubercart image field`" - I may think I'm done, so it's easy to miss this important additional message, which basically creates the field needed for image support. It's great to have the warning in the Status Report saying that there is a problem with Product images. But I'm wondering:

- Perhaps there is a way to clarify what the "Automatically configure products" message means (which BTW it's not so "automatic" since you actually have to click a link to do it). Perhaps change the wording to "Default image field not yet added to products! [link]Add default image field to products[link]". And also the Status Report could explain a bit more exactly what the problem is (it's not so much that "Ubercart product images are not yet configured", but rather that "Ubercart product **default image field** is not yet attached to the store products." or something of that nature.
- Instead of modifying the radio description to add this prompt, perhaps this could be done as a warning message? Or make the description really visible with red or something?

**Comment by bugfolder** [comment]
I agree that "Automatically configure..." is not so clear.

Let me back up a second to enumerate the use cases.

1. For people doing a clean install, the `uc_product_image` field will already have been created, so they don't need the warning visible.
2. If those people subsequently delete the `uc_product_image` field, the setting will revert to "No images needed or...", so they don't need the warning visible, either.
3. For people upgrading from a D7 site that didn't use the `uc_product_image` field (this was my experience that led me to this issue), they will see the warning on the Status Report page, which sends them to the Product Settings page; the radio button will be set to "Product image support provided by Ubercart...," and so they should definitely see the extra message.
4. If a site is currently set to "No images needed or..." and the person selects the Default..." radio button on the Product settings page, they'll see the extra message.

Now, case (4) could apply to someone who is just going "I wonder what this will do?", and so I don't see the need to put up a Backdrop warning message for them just because they clicked the button (they haven't saved anything, they could switch it back after they see the additional message). And having the extra message appear right on the thing that they just clicked (via the AJAX) makes sense to me, because that's the part of the page that they're already looking at.

For case (3), those people have just been sent to the Product Settings page from a warning message on the Status report, so I think they'll be looking for further guidance to be visible on the Product settings page. And fortunately, this tab is the topmost vertical tab, so it will be visible to them; in fact, it's the first thing on the page.

So, I'm thinking the best thing is just to change the extra text to something that's more blunt and straightforward, along the lines of what you suggested:

> Default image field not yet added to products! [link]Add default image field to products[link].

But I would suggest something slightly different for the additional text:

> The default image field needs to be added to products. [Add it now](#).

So that the full description text would be

> Product image support provided by Ubercart. The default image field needs to be added to products. [Add it now](#).

What do you think?

**Comment by argiepiano** [comment]
Hi @bugfolder. Thanks for the detailed explanation. I think the full description text with the "Add it now" looks good! 

After that adjustment LGTM!

**Comment by bugfolder** [comment]
Adjusted!

![image](https://user-images.githubusercontent.com/29214084/225636583-4c19db94-121e-4676-a349-0caec3841fe1.jpeg)

**Comment by argiepiano** [comment]
RTBC!
